File: modules/_foundation/src/main/java/com/projecthub/base/project/domain/validation/ProjectValidator.java
package com.projecthub.base.project.domain.validation;

import com.projecthub.base.project.api.dto.ProjectDTO;
import com.projecthub.base.project.infrastructure.repository.ProjectRepository;
import com.projecthub.base.shared.validation.ValidationResult;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Component
@RequiredArgsConstructor
public class ProjectValidator {
    
    private final ProjectRepository projectRepository;

    public ValidationResult validateCreate(ProjectDTO project) {
        List<String> errors = new ArrayList<>();

        validateName(project.name(), project.teamId(), errors);
        validateDates(project.startDate(), project.deadline(), errors);
        
        return new ValidationResult(errors);
    }

    public ValidationResult validateUpdate(UUID projectId, ProjectDTO project) {
        List<String> errors = new ArrayList<>();

        // Allow same name for same project ID
        validateNameForUpdate(projectId, project.name(), project.teamId(), errors);
        validateDates(project.startDate(), project.deadline(), errors);
        
        return new ValidationResult(errors);
    }

    private void validateName(String name, UUID teamId, List<String> errors) {
        if (projectRepository.existsByNameAndTeamId(name, teamId)) {
            errors.add("Project with this name already exists for the team");
        }
    }

    private void validateNameForUpdate(UUID projectId, String name, UUID teamId, List<String> errors) {
        if (projectRepository.existsByNameAndTeamId(name, teamId)) {
            projectRepository.findById(projectId).ifPresent(existingProject -> {
                if (!existingProject.getName().equals(name)) {
                    errors.add("Project with this name already exists for the team");
                }
            });
        }
    }

    private void validateDates(LocalDate startDate, LocalDate deadline, List<String> errors) {
        if (startDate != null && deadline != null && deadline.isBefore(startDate)) {
            errors.add("Deadline cannot be before start date");
        }
        
        if (deadline != null && deadline.isBefore(LocalDate.now())) {
            errors.add("Deadline cannot be in the past");
        }
    }
}

File: src/main/kotlin/com/projecthub/domain/project/validation/ProjectValidator.kt
package com.projecthub.domain.project.validation

import com.projecthub.domain.validation.ValidationResult
import com.projecthub.domain.validation.Validator
import com.projecthub.domain.project.Project
import com.projecthub.domain.project.ProjectRepository
import java.time.Instant
import java.time.temporal.ChronoUnit

/**
 * Domain validator for Project entities.
 * Implements core business rules for project validation including:
 * - Project name uniqueness within a team
 * - Date validations (start date before due date, due date not in past)
 */
class ProjectValidator(
    private val projectRepository: ProjectRepository
) : Validator<Project> {

    /**
     * Validates a project for creation
     */
    suspend fun validateCreate(project: Project): ValidationResult {
        return ValidationResult.Builder().apply {
            validateName(project.name, project.ownerId.value)
            validateDates(project.startDate, project.dueDate)
        }.build()
    }

    /**
     * Validates a project for update, allowing same name for same project
     */
    suspend fun validateUpdate(projectId: String, project: Project): ValidationResult {
        return ValidationResult.Builder().apply {
            validateNameForUpdate(projectId, project.name, project.ownerId.value)
            validateDates(project.startDate, project.dueDate)
        }.build()
    }

    private suspend fun ValidationResult.Builder.validateName(name: String, ownerId: String) {
        if (projectRepository.existsByNameAndTeamId(name, ownerId)) {
            addError("Project with this name already exists for the team")
        }
    }

    private suspend fun ValidationResult.Builder.validateNameForUpdate(
        projectId: String,
        name: String,
        ownerId: String
    ) {
        if (projectRepository.existsByNameAndTeamId(name, ownerId)) {
            projectRepository.findById(projectId)?.let { existingProject ->
                if (existingProject.name != name) {
                    addError("Project with this name already exists for the team")
                }
            }
        }
    }

    private fun ValidationResult.Builder.validateDates(
        startDate: Instant,
        dueDate: Instant?
    ) {
        dueDate?.let { deadline ->
            if (deadline.isBefore(startDate)) {
                addError("Due date cannot be before start date")
            }

            // Truncate to days for date-only comparison
            val today = Instant.now().truncatedTo(ChronoUnit.DAYS)
            val dueDateTruncated = deadline.truncatedTo(ChronoUnit.DAYS)

            if (dueDateTruncated.isBefore(today)) {
                addError("Due date cannot be in the past")
            }
        }
    }
}
